templates/rabbitmq.config.epp

```diff
@@ -25,6 +25,21 @@
     {cluster_nodes, {[], <%= $rabbitmq::config::cluster_node_type %>}},
     <%- } -%>
     {cluster_partition_handling, <%= $rabbitmq::config::cluster_partition_handling %>},
+<% } -%>
+<% unless $rabbitmq::config::quorum_membership_reconciliation_enabled =~ Undef {-%>
```

Comment:
Thanks for the assistance with this. Ill admit I'm not a big rubyist either. Yesterday I ended up down a big rabbithole (pun intended) and google marathon getting this to work and test cases to pass. At the end of the day, there are 3 specific cases I wanted to implement and cover idempotently:
Case 1: If undef, undefined, nil -> should literally "do nothing" - that is, there should be no entry in rabbitmq.config at all, and it will therefore revert to the rabbitmq global default setting (if the rabbitmq version you are on has one). This keeps this code "safe" and fully backward compatible with any RabbitMQ version
Case 2: If "true" -> should explicitly write to the rabbitmq.config file the true value
Case 3: If "false" -> should explicitly write to the rabbitmq.config file our desired false setting

My logic here was:
-All Vars are configured optional and to default to undef. This makes this "safe" in it can be applied even to a rabbitmq version that doesn't support these parameters, you simply cannot "set" them until you upgrade to a supported version, Global defaults for all these variables exist in rabbitmq 3.13+ - so they don't need to be set unless you "want to" I did make mention in the Reference.MD doc that these parameters require 3.13+ I was toying with adding a notify or failure message if the RabbitMQ version was below that, but was hesitant to due to potential upgrade race conditions. If someone on 3.12 "upgrades" and at the same time tries to input these params net-new on a single puppet run, the first puppetrun would still register the rabbitmq version "fact" at 3.12. It wouldn't register 3.13 until the subsequent run. This would cause a failure/notify. Notify might work as it wouldn't block, but I didn't see other examples in the code that led me to believe it was normal/standard for this repo.
- Apparently, through my research (not a ruby expert) i found that undef, undefined, and nil can all be valid "empty/unset" as far as EPP templates are concerned. I actually pulled the "unless" modifier verbatim from the puppet Documentation on EPP templates: https://puppet.com/docs/puppet/5.5/lang_template_epp.html and that "seems" like what they suggest for an optional variable.
- Simply checking for "if true" (my initial test here) didn't cover the undef and false cases idempotetly, and I wanted to cover that to allow explicitly setting true AND false (to be able to fully override a global default, should rabbitmq change the global default in the future.

Reply:
Ah, yes, you're right that for it to be an optional bool, you need the undef / true / false behavior for the booleans too. That makes sense.
Sorry for missing that the version requirement was in there

I think it makes sense to not check the fact. There are some version guards that check for puppet version, but they're mostly in the provider code, I think.

You're also right about the Puppet examples using `=~ Undef`... going to need to ask some folks about that, but I think either should work.

Reply:
Also, is there a clear case where setting the enabled parameter to false _explicitly_ is useful? If not, (and assuming the default is `false`), I think the best scenario for now might be to make it a boolean with a default of `false` vs. an optional bool.

Reply:
@wyardley - i think functionally defaulting to false, and having false "omit" from config altogether, with true adding to the config is functionally identical to my solution. The only case where it might cause an issue would be a specific hypothetical case in the future where RabbitMQ changes their default behavior for this to "true", In this case, i would no longer have a way to disable it, as flipping it to false would remove from the rabbitmq.config, inheriting the default to "true".
While that's a hypothetical future case, this is why i coded it to fully handle true/false/undef. However, if its not worth accounting for them ever changing the RabbitMQ default, i wouldn't be opposed to re-writing to be:
- Default False, if false, omit from rabbitmq.config and inherit global default (also preserves compatibility with RabbitMQ < 3.13)
if true - set to true in rabbitmq.config

I think the integer values, i likely would still need to use the undef / (integer) optional construct, as I'd need to have some form of undef/false to exclude it altogether for RabbitMQ < 3.13 to avoid breaking things.

Reply:
Yes, those other configuration parameters should remain optional, as well as the other bool probably.

I'm guessing it's unlikely for the value default to change in the future, but could be worth asking the RMQ folks if there's an easy way? I think usually we try to avoid optional bools where possible.

Reply:
Also @binford2k straightened me out here... in this context, `=~` is apparently a "type match operator", and not related to regex. So leave it as-is... maybe I'll go in and adjust the others in the file for consistency (separately) later.

Reply:
https://www.puppet.com/docs/puppet/8/lang_data_type#lang_data_type_usage-match-expressions

Docs on above ☝️